c/tskit/trees.c

```diff
@@ -2355,28 +2355,157 @@ compute_general_two_site_stat_result(const tsk_bit_array_t *site_a_state,
     return ret;
 }
 
+void
+stat_matrix_site_indicies_free(struct stat_matrix_site_indicies *idx)
+{
+    tsk_safe_free(idx->sites);
+    tsk_safe_free(idx->rshr);
+    tsk_safe_free(idx->cshr);
+    tsk_safe_free(idx->rdiff);
+    tsk_safe_free(idx->cdiff);
+    tsk_safe_free(idx->shr_idx);
+    tsk_safe_free(idx->rdiff_idx);
+    tsk_safe_free(idx->cdiff_idx);
+}
+
 static int
-get_mutation_samples(
-    const tsk_treeseq_t *ts, tsk_size_t *num_alleles, tsk_bit_array_t *allele_samples)
+realloc_and_set(tsk_size_t orig_size, tsk_size_t n, tsk_size_t *in, tsk_size_t **out)
+{
+    int ret = 0;
+    tsk_size_t *tmp = NULL;
+
+    if (n == orig_size) {
+        *out = in;
+        goto out;
+    }
+    if (n == 0) {
+        tsk_safe_free(in);
+        *out = NULL;
+        goto out;
+    }
+    tmp = tsk_realloc(in, n * sizeof(*tmp));
+    if (tmp == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    *out = tmp;
+out:
+    return ret;
+}
```

Comment:
This reallocation routine allows us to only store the individual sites we've specified, instead of arrays that are over-allocated to store all possible sites. This could be useful in large tree sequences.